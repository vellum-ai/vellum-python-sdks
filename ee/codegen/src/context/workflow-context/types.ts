export interface SDK_MODULE_PATHS {
  WORKFLOWS_MODULE_PATH: readonly string[];
  CORE_NODES_MODULE_PATH: readonly string[];
  DISPLAYABLE_NODES_MODULE_PATH: readonly string[];
  INPUTS_MODULE_PATH: readonly string[];
  SANDBOX_RUNNER_MODULE_PATH: readonly string[];
  STATE_MODULE_PATH: readonly string[];
  NODE_DISPLAY_MODULE_PATH: readonly string[];
  NODE_DISPLAY_TYPES_MODULE_PATH: readonly string[];
  VELLUM_TYPES_MODULE_PATH: readonly string[];
  PORTS_MODULE_PATH: readonly string[];
}
