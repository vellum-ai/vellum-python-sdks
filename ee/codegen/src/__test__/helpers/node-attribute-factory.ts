import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
