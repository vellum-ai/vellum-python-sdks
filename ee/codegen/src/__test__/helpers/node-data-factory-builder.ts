import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
