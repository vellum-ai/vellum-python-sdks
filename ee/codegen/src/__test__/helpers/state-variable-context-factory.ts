import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
