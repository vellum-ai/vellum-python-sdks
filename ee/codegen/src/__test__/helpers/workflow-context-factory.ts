import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
