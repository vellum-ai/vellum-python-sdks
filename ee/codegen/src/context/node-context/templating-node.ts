import { VellumVariableType } from "vellum-ai/api";

import { BaseNodeContext } from "src/context/node-context/base";
import { PortContext } from "src/context/port-context";
import { TemplatingNode } from "src/types/vellum";

export class TemplatingNodeContext extends BaseNodeContext<TemplatingNode> {
  baseNodeClassName = "TemplatingNode";
  baseNodeDisplayClassName = "BaseTemplatingNodeDisplay";
  isCore = true;

  public getNodeOutputNamesById(): Record<string, string> {
    const errorOutputId = this.getErrorOutputId();
    return {
      [this.nodeData.data.outputId]: "result",
      ...(errorOutputId ? { [errorOutputId]: "error" } : {}),
    };
  }

  protected getNodeOutputTypesById(): Record<string, VellumVariableType> {
    return {
      [this.nodeData.data.outputId]: this.nodeData.data.outputType,
      ...(this.nodeData.data.errorOutputId
        ? { [this.nodeData.data.errorOutputId]: "ERROR" }
        : {}),
    };
  }

  createPortContexts(): PortContext[] {
    return [
      new PortContext({
        workflowContext: this.workflowContext,
        nodeContext: this,
        portId: this.nodeData.data.sourceHandleId,
      }),
    ];
  }
}
