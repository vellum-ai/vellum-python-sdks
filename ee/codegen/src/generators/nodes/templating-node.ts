import { python } from "@fern-api/python-ast";
import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { VellumVariableType } from "vellum-ai/api/types";

import { OUTPUTS_CLASS_NAME, VELLUM_CLIENT_MODULE_PATH } from "src/constants";
import { TemplatingNodeContext } from "src/context/node-context/templating-node";
import { NodeAttributeGenerationError } from "src/generators/errors";
import { BaseNode } from "src/generators/nodes/bases/base";
import { TemplatingNode as TemplatingNodeType } from "src/types/vellum";
import { getVellumVariablePrimitiveType } from "src/utils/vellum-variables";

const TEMPLATING_INPUT_KEY = "template";
const INPUTS_PREFIX = "inputs";

export class TemplatingNode extends BaseNode<
  TemplatingNodeType,
  TemplatingNodeContext
> {
  protected getNodeAttributeNameByNodeInputKey(nodeInputKey: string): string {
    if (nodeInputKey === TEMPLATING_INPUT_KEY) {
      return nodeInputKey;
    }

    return `${INPUTS_PREFIX}.${nodeInputKey}`;
  }

  protected getNodeBaseGenericTypes(): AstNode[] {
    const stateType = this.getStateTypeOrBaseState();
    const primitiveOutputType = this.generateOutputType(
      this.nodeData.data.outputType
    );
    return [stateType, primitiveOutputType];
  }

  protected getNodeClassBodyStatements(): AstNode[] {
    const statements: AstNode[] = [];

    const otherInputs = Array.from(this.nodeInputsByKey.values()).filter(
      (nodeInput) =>
        nodeInput.nodeInputData.id !== this.nodeData.data.templateNodeInputId
    );

    statements.push(
      python.field({
        name: TEMPLATING_INPUT_KEY,
        initializer: this.getTemplatingInput(),
      })
    );

    statements.push(
      python.field({
        name: INPUTS_PREFIX,
        initializer: python.TypeInstantiation.dict(
          otherInputs.map((codeInput) => ({
            key: python.TypeInstantiation.str(codeInput.nodeInputData.key),
            value: codeInput,
          })),
          {
            endWithComma: true,
          }
        ),
      })
    );

    return statements;
  }

  protected getNodeDisplayClassBodyStatements(): AstNode[] {
    const statements: AstNode[] = [];

    statements.push(
      python.field({
        name: "target_handle_id",
        initializer: python.TypeInstantiation.uuid(
          this.nodeData.data.targetHandleId
        ),
      })
    );

    return statements;
  }

  private getTemplatingInput() {
    const templatingInput = this.nodeData.inputs.find(
      (input) => input.id === this.nodeData.data.templateNodeInputId
    );
    if (!templatingInput) {
      this.workflowContext.addError(
        new NodeAttributeGenerationError(
          "Templating input not found",
          "WARNING"
        )
      );
      return python.TypeInstantiation.str("");
    }

    const templateRule = templatingInput.value.rules[0];
    if (!templateRule) {
      this.workflowContext.addError(
        new NodeAttributeGenerationError(
          "Templating input rule not found",
          "WARNING"
        )
      );
      return python.TypeInstantiation.str("");
    }

    if (templateRule.type !== "CONSTANT_VALUE") {
      this.workflowContext.addError(
        new NodeAttributeGenerationError(
          "Templating input rule is not a constant value",
          "WARNING"
        )
      );
      return python.TypeInstantiation.str("");
    }

    if (templateRule.data.type !== "STRING") {
      this.workflowContext.addError(
        new NodeAttributeGenerationError(
          "Templating input rule is not a string",
          "WARNING"
        )
      );
      return python.TypeInstantiation.str("");
    }

    if (!templateRule.data.value) {
      this.workflowContext.addError(
        new NodeAttributeGenerationError(
          "Templating input rule value must be defined and nonempty",
          "WARNING"
        )
      );
      return python.TypeInstantiation.str("");
    }

    return python.TypeInstantiation.str(templateRule.data.value, {
      multiline: true,
      startOnNewLine: true,
      endWithNewLine: true,
    });
  }

  protected getOutputDisplay(): python.Field {
    return python.field({
      name: "output_display",
      initializer: python.TypeInstantiation.dict([
        {
          key: python.reference({
            name: this.nodeContext.nodeClassName,
            modulePath: this.nodeContext.nodeModulePath,
            attribute: [OUTPUTS_CLASS_NAME, "result"],
          }),
          value: python.instantiateClass({
            classReference: python.reference({
              name: "NodeOutputDisplay",
              modulePath:
                this.workflowContext.sdkModulePathNames
                  .NODE_DISPLAY_TYPES_MODULE_PATH,
            }),
            arguments_: [
              python.methodArgument({
                name: "id",
                value: python.TypeInstantiation.uuid(
                  this.nodeData.data.outputId
                ),
              }),
              python.methodArgument({
                name: "name",
                value: python.TypeInstantiation.str("result"),
              }),
            ],
          }),
        },
      ]),
    });
  }

  protected getErrorOutputId(): string | undefined {
    return this.nodeData.data.errorOutputId;
  }

  private generateOutputType(outputType: VellumVariableType): python.Type {
    return outputType === VellumVariableType.Json
      ? python.Type.reference(
          python.reference({
            name: "Json",
            modulePath: [
              ...VELLUM_CLIENT_MODULE_PATH,
              "workflows",
              "types",
              "core",
            ],
          })
        )
      : getVellumVariablePrimitiveType(outputType);
  }
}
