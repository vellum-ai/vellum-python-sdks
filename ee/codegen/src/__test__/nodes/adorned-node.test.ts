import { Writer } from "@fern-api/python-ast/core/Writer";
import { beforeEach } from "vitest";

import { workflowContextFactory } from "src/__test__/helpers";
import {
  retryAdornmentFactory,
  templatingNodeFactory,
  tryAdornmentFactory,
} from "src/__test__/helpers/node-data-factories";
import { createNodeContext, WorkflowContext } from "src/context";
import { TemplatingNodeContext } from "src/context/node-context/templating-node";
import { TemplatingNode } from "src/generators/nodes/templating-node";
import { AdornmentNode } from "src/types/vellum";

describe("Adorned Node", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;
  let node: TemplatingNode;

  beforeEach(() => {
    workflowContext = workflowContextFactory();
    writer = new Writer();
  });

  describe("try adornment no error code", () => {
    beforeEach(async () => {
      const adornments: AdornmentNode[] = [tryAdornmentFactory()];
      const nodeData = templatingNodeFactory()
        .withAdornments(adornments)
        .build();

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as TemplatingNodeContext;

      node = new TemplatingNode({
        workflowContext,
        nodeContext,
      });
    });

    it("getNodeFile", async () => {
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("try adornment with error code", () => {
    beforeEach(async () => {
      const adornments: AdornmentNode[] = [
        tryAdornmentFactory({ errorCode: "INVALID_CODE" }),
      ];
      const nodeData = templatingNodeFactory()
        .withAdornments(adornments)
        .build();

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as TemplatingNodeContext;

      node = new TemplatingNode({
        workflowContext,
        nodeContext,
      });
    });

    it("getNodeFile", async () => {
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("Retry adornment no error code", () => {
    beforeEach(async () => {
      const adornments: AdornmentNode[] = [retryAdornmentFactory()];
      const nodeData = templatingNodeFactory()
        .withAdornments(adornments)
        .build();

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as TemplatingNodeContext;

      node = new TemplatingNode({
        workflowContext,
        nodeContext,
      });
    });

    it("getNodeFile", async () => {
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("retry adornment with error code and max attempts and delay", () => {
    beforeEach(async () => {
      const adornments: AdornmentNode[] = [
        retryAdornmentFactory({
          errorCode: "INVALID_CODE",
          maxAttempts: 5,
          delay: 2.5,
        }),
      ];
      const nodeData = templatingNodeFactory()
        .withAdornments(adornments)
        .build();

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as TemplatingNodeContext;

      node = new TemplatingNode({
        workflowContext,
        nodeContext,
      });
    });

    it("getNodeFile", async () => {
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });
});
