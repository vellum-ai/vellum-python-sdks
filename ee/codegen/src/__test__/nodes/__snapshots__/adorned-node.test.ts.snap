// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Adorned Node > Retry adornment no error code > getNodeFile 1`] = `
"from vellum.workflows.nodes.core.retry_node.node import RetryNode
from vellum.workflows.nodes.displayable import TemplatingNode as BaseTemplatingNode
from vellum.workflows.state import BaseState


@RetryNode.wrap()
class TemplatingNode(BaseTemplatingNode[BaseState, str]):
    template = """Hello, World!"""
    inputs = {}
"
`;

exports[`Adorned Node > retry adornment with error code and max attempts and delay > getNodeFile 1`] = `
"from vellum.workflows.errors.types import WorkflowErrorCode
from vellum.workflows.nodes.core.retry_node.node import RetryNode
from vellum.workflows.nodes.displayable import TemplatingNode as BaseTemplatingNode
from vellum.workflows.state import BaseState


@RetryNode.wrap(
    retry_on_error_code=WorkflowErrorCode.INVALID_CODE, max_attempts=5, delay=2.5
)
class TemplatingNode(BaseTemplatingNode[BaseState, str]):
    template = """Hello, World!"""
    inputs = {}
"
`;

exports[`Adorned Node > try adornment no error code > getNodeFile 1`] = `
"from vellum.workflows.nodes.core.try_node.node import TryNode
from vellum.workflows.nodes.displayable import TemplatingNode as BaseTemplatingNode
from vellum.workflows.state import BaseState


@TryNode.wrap()
class TemplatingNode(BaseTemplatingNode[BaseState, str]):
    template = """Hello, World!"""
    inputs = {}
"
`;

exports[`Adorned Node > try adornment with error code > getNodeFile 1`] = `
"from vellum.workflows.errors.types import WorkflowErrorCode
from vellum.workflows.nodes.core.try_node.node import TryNode
from vellum.workflows.nodes.displayable import TemplatingNode as BaseTemplatingNode
from vellum.workflows.state import BaseState


@TryNode.wrap(on_error_code=WorkflowErrorCode.INVALID_CODE)
class TemplatingNode(BaseTemplatingNode[BaseState, str]):
    template = """Hello, World!"""
    inputs = {}
"
`;
